### ISSUE TITLE
Geth macOS can never pass 3899354

### ISSUE THREAD

**Issue by JC-000** [ethereum/go-ethereum#14954]
#### System information

Geth version: `geth version`
OS & Version: OSX 10.13 Beta (17A330h)
Commit hash : (if `develop`)

#### Expected behaviour

Geth synchronization should complete normally

#### Actual behaviour

Geth will never finish synchronizing, rejecting hash chains from this point forward.

#### Steps to reproduce the behaviour

Start fresh telling geth to do a fast sync.

#### Backtrace

INFO [08-09|15:45:17] Imported new block receipts              count=2    elapsed=6.323ms   number=3899349 hash=14b328…4c049c ignored=0
WARN [08-09|15:45:17] Rolled back headers                      count=368  header=3899229->3899229 fast=3899349->3899349 block=0->0
WARN [08-09|15:45:17] Synchronisation failed, dropping peer    peer=7ea8b54fd76d42ad err="retrieved hash chain is invalid"
WARN [08-09|15:45:22] Synchronisation failed, retrying         err="block download canceled (requested)"
INFO [08-09|15:46:50] Imported new state entries               count=1    flushed=0   elapsed=795µs     processed=18521315 pending=17    retry=0   duplicate=209 unexpected=1470
INFO [08-09|15:46:50] Imported new state entries               count=2    flushed=0   elapsed=1.282ms   processed=18521317 pending=49    retry=0   duplicate=209 unexpected=1470
INFO [08-09|15:46:55] Imported new state entries               count=3    flushed=0   elapsed=1.634ms   processed=18521320 pending=83    retry=14  duplicate=209 unexpected=1470
WARN [08-09|15:46:55] Rolled back headers                      count=760  header=3899229->3899229 fast=3899349->3899349 block=0->0
WARN [08-09|15:46:55] Synchronisation failed, dropping peer    peer=3688219a472fa5fc err="retrieved hash chain is invalid"
WARN [08-09|15:56:08] Synchronisation failed, dropping peer    peer=95e36d42acd396b8 err=timeout
WARN [08-09|15:57:13] Lied about chain head                    peer=957eabc288fba694 reported=1456791 found=1456967
WARN [08-09|15:57:13] Synchronisation failed, dropping peer    peer=957eabc288fba694 err="peer is stalling"
INFO [08-09|16:10:27] Imported new block headers               count=0    elapsed=3.075ms   number=3899541 hash=2da617…1acf1f ignored=192
INFO [08-09|16:10:27] Imported new block headers               count=0    elapsed=6.048ms   number=3899925 hash=69208b…2efaf2 ignored=384
WARN [08-09|16:10:27] Rolled back headers                      count=191  header=3899229->3899229 fast=3899349->3899349 block=0->0
WARN [08-09|16:10:27] Synchronisation failed, dropping peer    peer=ea4f678cc9c81721 err="retrieved hash chain is invalid"
INFO [08-09|16:10:36] Imported new state entries               count=1    flushed=0   elapsed=1.263ms   processed=18521321 pending=17    retry=0   duplicate=209 unexpected=1470
INFO [08-09|16:10:36] Imported new state entries               count=2    flushed=0   elapsed=2.048ms   processed=18521323 pending=49    retry=0   duplicate=209 unexpected=1470
INFO [08-09|16:10:37] Imported new block headers               count=0    elapsed=3.106ms   number=3899541 hash=2da617…1acf1f ignored=192
INFO [08-09|16:10:37] Imported new state entries               count=2    flushed=0   elapsed=4.081ms   processed=18521325 pending=80    retry=6   duplicate=209 unexpected=1470
WARN [08-09|16:10:37] Rolled back headers                      count=478  header=3899229->3899229 fast=3899349->3899349 block=0->0
WARN [08-09|16:10:37] Synchronisation failed, dropping peer    peer=f1263428cc2b1455 err="retrieved hash chain is invalid"
WARN [08-09|16:11:02] Ancestor below allowance                 peer=b21a4077f37eda11 number=2462989 hash=eecf25…6af862 allowance=3809349
WARN [08-09|16:11:02] Synchronisation failed, dropping peer    peer=b21a4077f37eda11 err="retrieved ancestor is invalid"
INFO [08-09|16:12:04] Imported new block headers               count=0    elapsed=3.066ms   number=3899541 hash=2da617…1acf1f ignored=192
INFO [08-09|16:12:05] Imported new block headers               count=0    elapsed=3.141ms   number=3899733 hash=0c2b79…6ea507 ignored=192
INFO [08-09|16:12:06] Imported new block receipts              count=3    elapsed=4.838ms   number=3899352 hash=19cb8b…2d9496 ignored=0
INFO [08-09|16:12:07] Imported new block receipts              count=2    elapsed=4.197ms   number=3899354 hash=367eff…610e3f ignored=0
WARN [08-09|16:12:07] Rolled back headers                      count=291  header=3899229->3899229 fast=3899354->3899354 block=0->0
WARN [08-09|16:12:07] Synchronisation failed, dropping peer    peer=1708ab6bffe942f1 err="retrieved hash chain is invalid"
INFO [08-09|16:17:03] Imported new block headers               count=0    elapsed=3.09ms    number=3899546 hash=a69df7…68bdc2 ignored=192
WARN [08-09|16:17:05] Rolled back headers                      count=575  header=3899229->3899229 fast=3899354->3899354 block=0->0
WARN [08-09|16:17:05] Synchronisation failed, dropping peer    peer=8f568ea7154bee3f err="retrieved hash chain is invalid"
WARN [08-09|16:30:59] Synchronisation failed, dropping peer    peer=9848b2a90b6225ff err="action from bad peer ignored"
INFO [08-09|16:40:03] Imported new state entries               count=1    flushed=0   elapsed=1.285ms   processed=18521326 pending=17    retry=0   duplicate=209 unexpected=1470
INFO [08-09|16:40:03] Imported new state entries               count=2    flushed=0   elapsed=3.189ms   processed=18521328 pending=49    retry=0   duplicate=209 unexpected=1470
INFO [08-09|16:40:04] Imported new state entries               count=4    flushed=0   elapsed=9.134ms   processed=18521332 pending=113   retry=0   duplicate=209 unexpected=1470
INFO [08-09|16:40:04] Imported new state entries               count=6    flushed=0   elapsed=12.205ms  processed=18521338 pending=135   retry=4   duplicate=209 unexpected=1470
INFO [08-09|16:40:05] Imported new state entries               count=8    flushed=0   elapsed=28.624ms  processed=18521346 pending=204   retry=4   duplicate=209 unexpected=1470
INFO [08-09|16:40:05] Imported new state entries               count=1    flushed=0   elapsed=5.375ms   processed=18521347 pending=205   retry=1   duplicate=209 unexpected=1470
INFO [08-09|16:40:05] Imported new block headers               count=0    elapsed=5.937ms   number=3899738 hash=995dd5…883a86 ignored=384
INFO [08-09|16:40:05] Imported new state entries               count=1    flushed=0   elapsed=1.831ms   processed=18521348 pending=206   retry=5   duplicate=209 unexpected=1470
INFO [08-09|16:40:05] Imported new state entries               count=1    flushed=0   elapsed=170µs     processed=18521349 pending=208   retry=7   duplicate=209 unexpected=1470
INFO [08-09|16:40:05] Imported new state entries               count=10   flushed=0   elapsed=24.494ms  processed=18521359 pending=278   retry=4   duplicate=209 unexpected=1470
INFO [08-09|16:40:06] Imported new state entries               count=12   flushed=9   elapsed=14.523ms  processed=18521371 pending=338   retry=5   duplicate=209 unexpected=1470
INFO [08-09|16:40:06] Imported new state entries               count=1    flushed=0   elapsed=3.331ms   processed=18521372 pending=340   retry=3   duplicate=209 unexpected=1470
WARN [08-09|16:40:06] Rolled back headers                      count=284  header=3899229->3899229 fast=3899354->3899354 block=0->0
WARN [08-09|16:40:06] Synchronisation failed, dropping peer    peer=f9139b9db3d61da0 err="retrieved hash chain is invalid"
INFO [08-09|16:47:39] Imported new state entries               count=1    flushed=0   elapsed=1.431ms   processed=18521373 pending=17    retry=0   duplicate=209 unexpected=1470
INFO [08-09|16:47:40] Imported new block headers               count=0    elapsed=3.13ms    number=3899546 hash=a69df7…68bdc2 ignored=192
INFO [08-09|16:47:42] Imported new block receipts              count=1    elapsed=404µs     number=3899355 hash=d36e67…a65d61 ignored=0
WARN [08-09|16:47:43] Rolled back headers                      count=575  header=3899229->3899229 fast=3899355->3899355 block=0->0
WARN [08-09|16:47:43] Synchronisation failed, dropping peer    peer=5796df28510ea126 err="retrieved hash chain is invalid"
INFO [08-09|16:51:23] Imported new state entries               count=1    flushed=0   elapsed=3.699ms   processed=18521374 pending=17    retry=0   duplicate=209 unexpected=1470
INFO [08-09|16:51:23] Imported new state entries               count=2    flushed=0   elapsed=11.009ms  processed=18521376 pending=49    retry=0   duplicate=209 unexpected=1470
INFO [08-09|16:51:24] Imported new state entries               count=3    flushed=0   elapsed=9.66ms    processed=18521379 pending=95    retry=2   duplicate=209 unexpected=1470
INFO [08-09|16:51:24] Imported new block headers               count=0    elapsed=3.274ms   number=3899547 hash=ae3a85…b80817 ignored=192
INFO [08-09|16:51:24] Imported new block headers               count=0    elapsed=3.752ms   number=3899739 hash=25cd5b…ade30c ignored=192
INFO [08-09|16:51:24] Imported new state entries               count=7    flushed=0   elapsed=23.209ms  processed=18521386 pending=138   retry=2   duplicate=209 unexpected=1470
INFO [08-09|16:51:24] Imported new state entries               count=15   flushed=1   elapsed=56.123ms  processed=18521401 pending=190   retry=2   duplicate=209 unexpected=1470
INFO [08-09|16:51:24] Imported new state entries               count=30   flushed=10  elapsed=75.326ms  processed=18521431 pending=269   retry=0   duplicate=209 unexpected=1470
INFO [08-09|16:51:25] Imported new block headers               count=0    elapsed=3.071ms   number=3899931 hash=c75402…a57d56 ignored=192
INFO [08-09|16:51:25] Imported new state entries               count=54   flushed=20  elapsed=174.116ms processed=18521485 pending=303   retry=6   duplicate=209 unexpected=1470
WARN [08-09|16:51:25] Rolled back headers                      count=111  header=3899229->3899229 fast=3899355->3899355 block=0->0
WARN [08-09|16:51:25] Synchronisation failed, dropping peer    peer=3f7c304fdbe4b09e err="retrieved hash chain is invalid"

````
[backtrace]
````

**Comment by lexab65** [comment]
Works fine for me.
command run:
$ cd ~/geth-darwin-amd64-1.6.7-ab5646c5/
$ ./geth

Geth/v1.6.7
MacOS High Sierra Version 10.13 Beta (17A330h)

**Comment by JC-000** [comment]
300GB of free space.  Mac Pro 2013/16GB/1TB/D500 FWIW.
On every start, Ethereum wallet warns me about clock sync, but it is synced.  I have tried toggling the system preference without effect, but I'll experiment some more.
I have nuked the ~/Library/Application Support/Ethereum/geth/chaindata folder and restarted a couple times without a change.  I have just removed the entire ~/Library/Application Support/Ethereum/ and restarted again because I think I might know what is happening.  It is at block ~780000 now.

**Comment by JC-000** [comment]
Ok it stalled for 45 minutes at block 4140932, so I restarted it a few times and on the third attempt it finally finished up.  Compared to the Windows 10 x64 machine sitting right next to it on the desk it took an incredibly long time.

As for it being stuck at 3899229, I think I know what happened.  After I started it up and let it sync the first time, and populated 0.25ETH into the wallet that I generated, I put a small transaction into place from one wallet address into another.  At the time I did it, I think geth was stalled, or had lost its internet connectivity.  I believe I quit and restarted after a while.  My guess is that  it would never accept 3899229 from other nodes because it was missing that 0.25ETH transaction that it expected.  By nuking the entire ~/Library/Ethereum/, that road block got cleared.  So perhaps there is either a transaction state management or garbage collection issue.
